pkg/kubelet/cm/memorymanager/policy_static.go

```diff
@@ -146,41 +162,88 @@ func (p *staticPolicy) Allocate(s state.State, pod *v1.Pod, container *v1.Contai
+		}
+
 		// Update nodes memory state
-		for _, nodeID := range maskBits {
-			machineState[nodeID].NumberOfAssignments++
-			machineState[nodeID].Cells = maskBits
+		p.updateMachineState(machineState, maskBits, resourceName, requestedSize)
+	}
 
-			// we need to continue to update all affinity mask nodes
-			if requestedSize == 0 {
-				continue
-			}
+	// save the old machine state before update it
+	p.updateStateToReuse(s, pod, container, containerBlocks)
+	p.updateInitContainersReusableMemory(pod, container, containerBlocks)
 
-			// update the node memory state
-			nodeResourceMemoryState := machineState[nodeID].MemoryMap[resourceName]
-			if nodeResourceMemoryState.Free <= 0 {
-				continue
-			}
+	s.SetMachineState(machineState)
+	s.SetMemoryBlocks(podUID, container.Name, containerBlocks)
 
-			// the node has enough memory to satisfy the request
-			if nodeResourceMemoryState.Free >= requestedSize {
-				nodeResourceMemoryState.Reserved += requestedSize
-				nodeResourceMemoryState.Free -= requestedSize
-				requestedSize = 0
-				continue
-			}
+	// update init containers memory requests under the state file
+	for _, containersReusableResources := range p.initContainersReusableMemory[podUID] {
+		for containerName, containerReusableResources := range containersReusableResources {
+			s.SetMemoryBlocks(podUID, containerName, containerReusableResources)
+		}
+	}
+
+	return nil
+}
+
+func (p *staticPolicy) updateMachineState(machineState state.NUMANodeMap, numaAffinity []int, resourceName v1.ResourceName, requestedSize uint64) {
+	for _, nodeID := range numaAffinity {
+		machineState[nodeID].NumberOfAssignments++
+		machineState[nodeID].Cells = numaAffinity
+
+		// we need to continue to update all affinity mask nodes
+		if requestedSize == 0 {
+			continue
+		}
+
+		// update the node memory state
+		nodeResourceMemoryState := machineState[nodeID].MemoryMap[resourceName]
+		if nodeResourceMemoryState.Free <= 0 {
+			continue
+		}
 
-			// the node does not have enough memory, use the node remaining memory and move to the next node
-			requestedSize -= nodeResourceMemoryState.Free
-			nodeResourceMemoryState.Reserved += nodeResourceMemoryState.Free
-			nodeResourceMemoryState.Free = 0
+		// the node has enough memory to satisfy the request
+		if nodeResourceMemoryState.Free >= requestedSize {
+			nodeResourceMemoryState.Reserved += requestedSize
+			nodeResourceMemoryState.Free -= requestedSize
+			requestedSize = 0
+			continue
 		}
+
+		// the node does not have enough memory, use the node remaining memory and move to the next node
+		requestedSize -= nodeResourceMemoryState.Free
+		nodeResourceMemoryState.Reserved += nodeResourceMemoryState.Free
+		nodeResourceMemoryState.Free = 0
 	}
+}
 
-	s.SetMachineState(machineState)
-	s.SetMemoryBlocks(string(pod.UID), container.Name, containerBlocks)
+func (p *staticPolicy) getPodReusableMemory(pod *v1.Pod, numaAffinity []int, resourceName v1.ResourceName) uint64 {
```

Comment:
This function makes alot of sense.  It walks through all blocks in all `init` containers from `p.initContainersReusableMemory[string(pod.UID)]` and if those blocks match the current resource type, it returns the amount of memory that can be reused from those blocks.

I assume an update to `p.initContainersReusableMemory[string(pod.UID)][containerName]` is made in a call just after this to account for the memory being pulled from it? Does it make sense to just do it inline or are there conditions where we would make this call to see how much we can reuse, but then failout before actually allocating it?

Reply:
The logic is:
- Hints generated when the re-usable memory taken into the account 
- Container blocks generated when the re-usable memory taken into the account
- The re-usable memory updated with generated container blocks

Because we call `GetTopologyHints` and `Allocate` sequentially for each container our calculations aligned and we have synced values between getting and updating the re-usable memory.

Reply:
> Does it make sense to just do it inline or are there conditions where we would make this call to see how much we can reuse, but then failout before actually allocating it?

Can you please clarify this sentence?

Reply:
I'm asking if it makes sense to just do the logic for part (2) of https://github.com/kubernetes/kubernetes/pull/99640/files#r659842084 inline in this function.

That way we both calculate how much memory to reuse as well as subtract it out from the proper `init` container's memory blocks all at the same time.

Reply:
The problem that we calling `getPodReusableMemory` also from the `GetTopologyHints`, and it should not update the re-usable memory because it can create some inconsistency if it will be some error between the call to `GetTopologyHints` and `Allocate`. I will try to see if it possible to move some common code into a separate method.

Reply:
I can move
```
if resourceName != containerBlock.Type {
                                        continue
                                }

                                if containerBlock.Size == 0 {
                                        continue
                                }

                                if !isNUMAAffinitiesEqual(numaAffinity, containerBlock.NUMAAffinity) {
                                        continue
                                }
```
to the separate method, but IMHO it not a big improvement

Reply:
Let me take another pass tomorrow before changing anything then.

Reply:
Yes. After reviewing the `CPUManager` code for this, I remember/understand why this separation is necessary.